File: graph/topological_sort/java/TopologicalSort.java
import java.util.*;
import org.apache.commons.lang3.ArrayUtils;

public class TopologicalSort {
  // TopologicalSort takes a DAG in adjacency list format, and output
  // the topological order.
  public static int[] sort(int[][] graph) {
    int n = graph.length;
    boolean[] visited = new boolean[n];
    List<Integer> l = new ArrayList<Integer>();
    for (int u = 0; u < n; u++) {
      if (!visited[u]) {
        explore(graph, u, visited, l);
      }
    }
    Collections.reverse(l);
    return ArrayUtils.toPrimitive(l.toArray(new Integer[l.size()]));
  }

  private static void explore(int[][] graph, int u, boolean[] visited, List<Integer> l) {
    visited[u] = true;
    for (int v : graph[u]) {
      if (!visited[v]) {
        explore(graph, v, visited, l);
      }
    }
    l.add(u);
  }
}

File: graph/topological_sort/py/topological_sort.py
# Topological sort takes a DAG in adjacency list formant, and out put
# the topological order.
#
# g: graph in adjacency list
# return: list to hold vertices in topological order
def sort(g):
    visited = {}
    for u in g:
        visited[u] = False
    l = []
    for u in g:
        if not visited[u]:
            explore(g, u, visited, l)
    return list(reversed(l))

def explore(g, u, visited, l):
    visited[u] = True
    for v in g[u]:
        if not visited[v]:
            explore(g, v, visited, l)
    l.append(u)
